File: experimental/webgtt/javascript/webgtt.js
// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file.

/**
 * @fileoverview This file contains the top level JavaScript required for the
 * WebGTT application. The two main functions are fired when the page loads and
 * when the NaCl module loads, respectively. We handle these load events
 * separately since the order in which these events are fired is much less
 * predictable than other standard DOM elements (for which we assume that they
 * are loaded when the page is loaded).
 *
 * @author ragad@google.com (Raga Gopalakrishnan)
 */

/**
 * Each of these variables is an object corresponding to a DOM element. Since
 * event handlers for one DOM element may need to modify properties of or
 * perform actions on another DOM element, these objects have global scope.
 */
var naclModule1 = undefined; // NaClModule object. See naclmodule.js
var graph1 = new Graph(); // Graph object. See graph.js
var canvas1 = undefined; // Canvas object. See canvas.js
var coloringButton1 = undefined; // ColoringButton object. See button.js

/**
 * This constant holds the default status message.
 * @type {string}
 * @const
 */
STATUS_TEXT = 'NO-STATUS';

/**
 * This function is triggered when the page has finished loading, and contains
 * starter code (initializing objects, etc.).
 *
 * The NaCl module could still be loading while the page has finished loading.
 * So, the status must be updated accordingly, and no NaClModule object is
 * created yet. New objects of the Canvas and ColoringButton classes are
 * created, and the appropriate event handlers are attached.
 */
function main1 () {
  if (naclModule1 == undefined) {
    updateStatus('LOADING...');
  } else {
    updateStatus();
  }

  canvas1 = new Canvas(document.getElementById('mainCanvas'), graph1);
  canvas1.drawingCanvas.addEventListener(
      'click', function (e) {canvas1.handleCanvasClick(e);}, false);
  window.addEventListener(
      'keydown', function (e) {canvas1.handleCanvasKeydown(e);}, true);

  coloringButton1 = new ColoringButton(document.getElementById('coloring'),
                                       graph1);
  coloringButton1.coloringButton.addEventListener(
      'click', function (e) {coloringButton1.handleColoringButtonClick(e);},
      false);
}

/**
 * This function is triggered when the NaCl module has finished loading. It
 * creates a new NaClModule object, attaches an event handler to it, and
 * indcates to the user that the NaCl module has loaded successfully.
 */
function main2() {
  naclModule1 = new NaClModule(document.getElementById('webgtt'), graph1);
  naclModule1.naclModule.addEventListener(
      'message', function (e) {naclModule1.handleMessage(e);}, false);
  updateStatus('SUCCESS');
}

/**
 * This function sets the global status message.
 *
 * If the element with id 'statusField' exists, then its HTML is set to the
 * status message as well. opt_message is the message text. If this is null or
 * undefined, then the element with id 'statusField' is set to STATUS_TEXT.
 *
 * @param {string} opt_message The status message to be displayed.
 */
function updateStatus(opt_message) {
  var statusText = STATUS_TEXT;
  if (opt_message) {
    statusText = opt_message;
  }
  var statusField = document.getElementById('status_field');
  if (statusField) {
    statusField.innerHTML = statusText;
  }
}

File: experimental/webgtt/webgtt.cc
// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file.

/// @fileoverview This file contains code for a simple prototype of the webgtt
/// project. it demonstrates loading, running and scripting a simple nacl
/// module, which, when given the adjacency matrix of a graph by the browser,
/// returns a valid vertex coloring of the graph. to load the nacl module, the
/// browser first looks for the createmodule() factory method. it calls
/// createmodule() once to load the module code from the .nexe. after the .nexe
/// code is loaded, createmodule() is not called again. once the .nexe code is
/// loaded, the browser then calls the createinstance() method on the object
/// returned by createmodule(). it calls createinstance() each time it
/// encounters an <embed> tag that references the nacl module.
///
/// @author ragad@google.com (Raga Gopalakrishnan)

#include <cmath>
#include <string>

#include "webgtt/parser.h"
#include "ppapi/cpp/instance.h"
#include "ppapi/cpp/module.h"
#include "ppapi/cpp/var.h"

namespace webgtt {

/// The Instance class. One of these exists for each instance of the NaCl module
/// on the web page. The browser will ask the Module object to create a new
/// Instance for each occurence of the <embed> tag that has these attributes:
///     type="application/x-nacl"
///     src="webgtt.nmf"
/// To communicate with the browser, the HandleMessage() method is overridden
/// for receiving messages from the browser. The PostMessage() method is used to
/// send messages back to the browser. This interface is entirely asynchronous.
class WebgttInstance : public pp::Instance {
 public:
  /// The constructor creates the plugin-side instance.
  ///
  /// @param[in] instance The handle to the browser-side plugin instance.
  /// @constructor
  explicit WebgttInstance(PP_Instance instance) : pp::Instance(instance) {}
  virtual ~WebgttInstance() {}

  /// This function handles messages coming in from the browser via
  /// postMessage().
  ///
  /// The @a var_message can contain anything: a JSON string, a string that
  /// encodes method names and arguments, etc.
  ///
  /// @param[in] var_message The message posted by the browser.
  virtual void HandleMessage(const pp::Var& var_message) {
    if (!var_message.is_string()) {
      return;
    }
    std::string message = var_message.AsString();

    Parser parse_message(message);
    pp::Var var_reply;
    if (parse_message.DecodeMessage()) {
      var_reply = pp::Var(parse_message.GetResponse());
    } else {
      var_reply = pp::Var("Error encountered while parsing the message!");
    }
    PostMessage(var_reply);
  }

 private:
  /// This disallows usage of copy and assignment constructors.
  WebgttInstance(const WebgttInstance&);
  void operator=(const WebgttInstance&);
};

/// The Module class. The browser calls the CreateInstance() method to create
/// an instance of the NaCl module on the web page. The browser creates a new
/// instance for each <embed> tag with type="application/x-nacl".
class WebgttModule : public pp::Module {
 public:
  WebgttModule() : pp::Module() {}
  virtual ~WebgttModule() {}

  /// This function creates and returns a WebgttInstance object.
  ///
  /// @param[in] instance The browser-side instance.
  /// @return The plugin-side instance.
  virtual pp::Instance* CreateInstance(PP_Instance instance) {
    return new WebgttInstance(instance);
  }

 private:
  /// This disallows usage of copy and assignment constructors.
  WebgttModule(const WebgttModule&);
  void operator=(const WebgttModule&);
};

}  // namespace webgtt

namespace pp {

/// This is the factory function called by the browser when the module is
/// first loaded. The browser keeps a singleton of this module. It calls the
/// CreateInstance() method on the object that is returned to make instances.
/// There is one instance per <embed> tag on the page. This is the main
/// binding point for the NaCl module with the browser.
Module* CreateModule() {
  return new webgtt::WebgttModule();
}

}  // namespace pp
